dbms/src/Common/HashTable/StringHashTable.h

```diff
@@ -0,0 +1,494 @@
+#pragma once
+
+#include <Common/HashTable/HashMap.h>
+#include <Common/HashTable/HashTable.h>
+
+/// TODO feature macros
+
+#define CASE_1_8 \
+    case 1: \
+    case 2: \
+    case 3: \
+    case 4: \
+    case 5: \
+    case 6: \
+    case 7: \
+    case 8
+
+#define CASE_9_16 \
+    case 9: \
+    case 10: \
+    case 11: \
+    case 12: \
+    case 13: \
+    case 14: \
+    case 15: \
+    case 16
+
+#define CASE_17_24 \
+    case 17: \
+    case 18: \
+    case 19: \
+    case 20: \
+    case 21: \
+    case 22: \
+    case 23: \
+    case 24
+
+struct StringKey0
+{
+};
+
+using StringKey8 = UInt64;
+using StringKey16 = DB::UInt128;
+struct StringKey24
+{
+#if !__clang__
+#    pragma GCC diagnostic push
+#    pragma GCC diagnostic ignored "-Wmaybe-uninitialized"
+#endif
+
+    UInt64 a;
+    UInt64 b;
+    UInt64 c;
+
+    bool operator==(const StringKey24 rhs) const { return a == rhs.a && b == rhs.b && c == rhs.c; }
+    bool operator!=(const StringKey24 rhs) const { return !operator==(rhs); }
+    bool operator==(const UInt64 rhs) const { return a == rhs && b == 0 && c == 0; }
+    bool operator!=(const UInt64 rhs) const { return !operator==(rhs); }
+
+#if !__clang__
+#    pragma GCC diagnostic pop
+#endif
+
+    StringKey24 & operator=(const UInt64 rhs)
+    {
+        a = rhs;
+        b = 0;
+        c = 0;
+        return *this;
+    }
+};
+
+struct ToStringRef
+{
```

Comment:
Why do we use this struct and not plain conversion functions (e.g. `StringRef toStringRef(const StringKey8&)`)?

Reply:
Hmm, it's used to be a template argument like Hash Trait. I'll turn them into overloads.